godot-core/src/engine/translate.rs

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+//! Macros for translation.
+
+pub use crate::{tr, tr_n};
+
+/// A convenience macro for using the [`Object::tr()`](crate::engine::Object::tr()) and [`Object::tr_ex()`](crate::engine::Object::tr_ex())
+///  methods.
+///
+/// The methods are called from the [`Engine`](crate::engine::Engine) singleton.
+///
+/// Takes a format string literal, with optional arguments. Optionally, `context` for potentially ambiguous words can be
+/// added before the format arguments, separated with a `;`.
+///
+/// For maximum flexibility and minimum ambiguity, prefer inlined, positional, or named arguments in the format strings:
+/// ```no_run
+/// # use godot_core::tr;
+/// # struct Foo { x: i32, y: i32 };
+/// # impl std::fmt::Display for Foo {
+/// #    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+/// #        write!(f, "")
+/// #    }
+/// # }
+/// # let a = Foo { x: 0, y: 0 };
+/// # let b = Foo { x: 0, y: 0 };
+/// # let context = "context";
+/// // Good
+/// tr!(context; "{a} is a {b}"); // inlined, with context
+/// tr!("{0} is a {1}", a, b); // positional, without context
+/// tr!("{c} is a {d}", c = a.x, d = b.y); // named (inlining not possible here)
+///
+/// // Not so good, much more fragile
+/// // and does not easily allow ordering arguments differently in other translations.
+/// tr!("{} is a {}", a, b);
+/// ```
+/// see also: <https://docs.godotengine.org/en/stable/tutorials/i18n/internationalizing_games.html#translation-contexts>
```

Comment:
I'd move the sentence about `Engine` down here, since it's a borderline implementation detail. It's also fine to 🪓 it, up to you. But it's not really required to know when using the macros.

```suggestion
/// The methods are called from the [`Engine`](crate::engine::Engine) singleton.
/// 
/// See also [Translation contexts](https://docs.godotengine.org/en/stable/tutorials/i18n/internationalizing_games.html#translation-contexts) in Godot.
```

Same for the other macro 🙂

Reply:
I think I'll leave it in, because it's not intuitively obvious that it's a method, then users don't have to dig deeper to figure out "wait if this has to be called from an instance, am I instantiating an object every time I use it?" I did debate taking it out altogether though, maybe once the API is stabilized and source is linked from the docs it could be cut, since yeah it's a bit wordy as-is.

Reply:
Sounds reasonable! As long as less important details are at the end, I don't think wordy is a big issue 😀

Reply:
You forgot the "See also" part. If I add a code suggestion, you can typically apply it directly on the web UI and then combine the commits later...

Reply:
Skimmed over it, my brain just saw a link, sorry.

Thanks for all the patience btw. If it's not obvious, most of my history with git has been on my own work where I only had to make things clear to myself and didn't have to worry about a clean commit history.